webrisk/database.go

```diff
@@ -0,0 +1,543 @@
+// leeway before declaring the database as stale.
+const (
+	maxRetryDelay  = 24 * time.Hour
+	baseRetryDelay = 15 * time.Minute
+	jitter         = 30 * time.Second
+)
+
+// database tracks the state of the threat lists published by the Webrisk API.
+// Since the global blacklist is constantly changing, the contents of the
+// database needs to be periodically synced with the Webrisk servers in
+// order to provide protection for the latest threats.
+//
+// The process for updating the database is as follows:
+//	* At startup, if a database file is provided, then load it. If loaded
+//	properly (not corrupted and not stale), then set tfu as the contents.
+//	Otherwise, pull a new threat list from the Web Risk API.
+//	* Periodically, synchronize the database with the Web Risk API.
+//	This uses the Version Token fields to update only parts of the threat list that have
+//	changed since the last sync.
+//	* Anytime tfu is updated, generate a new tfl.
+//
+// The process for querying the database is as follows:
+//	* Check if the requested full hash matches any partial hash in tfl.
+//	If a match is found, return a set of ThreatTypes with a partial match.
+type database struct {
+	config *Config
+
+	// threatsForUpdate maps ThreatTypes to lists of partial hashes.
+	// This data structure is in a format that is easily updated by the API.
+	// It is also the form that is written to disk.
+	tfu threatsForUpdate
+	mu  sync.Mutex // Protects tfu
+
+	// threatsForLookup maps ThreatTypes to sets of partial hashes.
+	// This data structure is in a format that is easily queried.
+	tfl threatsForLookup
+	ml  sync.RWMutex // Protects tfl, err, and last
+
+	err             error         // Last error encountered
+	readyCh         chan struct{} // Used for waiting until not in an error state.
+	last            time.Time     // Last time the threat list were synced
+	updateAPIErrors uint          // Number of times we attempted to contact the api and failed
+
+	log *log.Logger
+}
+
+type threatsForUpdate map[ThreatType]partialHashes
+type partialHashes struct {
+	// Since the Hashes field is only needed when storing to disk and when
+	// updating, this field is cleared except for when it is in use.
+	// This is done to reduce memory usage as the contents of this can be
+	// regenerated from the tfl.
+	Hashes hashPrefixes
+
+	SHA256 []byte // The SHA256 over Hashes
+	State  []byte // Arbitrary binary blob to synchronize state with API
+}
+
+type threatsForLookup map[ThreatType]hashSet
+
+// databaseFormat is a light struct used only for gob encoding and decoding.
+// As written to disk, the format of the database file is basically the gzip
+// compressed version of the gob encoding of databaseFormat.
+type databaseFormat struct {
+	Table threatsForUpdate
+	Time  time.Time
+}
+
+// Init initializes the database from the specified file in config.DBPath.
+// It reports true if the database was successfully loaded.
```

Comment:
Comment to check the db error (with `Status`?) if it returns false?

Reply:
Done.